### DISCUSSION TITLE
How to define an integral based function as the material properties

### DISCUSSION THREAD

**Discussion by sSajjad90** [idaholab/moose#16852] [Q&A Modules: General] [answered]
Hi all,
I will implement a code to calculate microstructure parameters associated with the precipitation and solute growth and then use the computed parameters to evaluate mechanical properties. One of those parameters (a1) is calculated in the attached equation for the given C* at each time step (t). Assume the other parameters ( C, C*beta, D,e12, and e13) are constant. 

I was wondering there is any material class as same as the "DerivativeParsedMaterial" function to implement an integral base function (like this) to calculate parameters in the material block? Or I have to define user objects to compute this function? If so, is there any suggestion or example?

Thanks in advanced 

![EQU](https://user-images.githubusercontent.com/78385504/106624338-8d229480-6543-11eb-981b-653017596d89.png)

**Comment by GiudGiud** [comment, marked as answer]
Hello

Are you looking to parse this function from the input file? I dont know if FParser can parse that. @roystgnr will know more. 
I would recommend computing it in your own material class, and just supply the parameters / variables in the input. 

Depending on what a1 is, you may also be able to define this as a new Function and call it from a FunctionMaterial. 
But this is only worth the effort if you are using this function in other places. 

Best,
Guillaume

**Reply by roystgnr** [reply]
It's possible @dschwen has added something (IIRC he was responsible for the differentiation capabilities in our FParser fork) for integrals, but I don't recall seeing anything of the sort.

**Comment by sSajjad90** [comment]
To explain a bit more...

 In fact, C* and a1 are unknown material properties that changed over time and are the input parameters for the primary analysis. I will solve C* in a separate function; then, it will be used to compute a1 in the attached equation for each time step and then update it in the next step. My difficulty is to find a relevant function to define this equation (includes an integral). 
I do not know any predefined class function to solve integrals (like "DerivativeParsedMaterial." using for the differentiation of the function expression) to calculate this equation.

many thanks,
Sajjad 
![c star](https://user-images.githubusercontent.com/78385504/106784165-08ef1080-661a-11eb-98c2-8c63b7c0d784.png)
![EQU](https://user-images.githubusercontent.com/78385504/106784197-10aeb500-661a-11eb-8c60-fccc750fdaa3.png)

**Comment by GiudGiud** [comment]
Hello
Did you solve this?
I think you may have to add the integration yourself. There are plenty of numerical integration techniques that will compute this integral to your desired accuracy. 

We would welcome the contribution as a general utility to integrate functions. See framework/src/utils for examples of these utility functions. 

Guillaume
